File: examples/python/01_parse_and_rewrite.py
"""Structural parse, search, and edit with the vorpal_py AST API.

    pip install vorpal-py
"""
from vorpal_py import SgRoot

code = """
def fetch(url):
    print("fetching", url)
    data = requests.get(url)
    print(data.status_code)
    return data
"""

root = SgRoot(code, "python").root()

# Pattern search with metavariable captures.
for call in root.find_all(pattern="print($$$ARGS)"):
    r = call.range()
    print(f"print at line {r.start.line + 1}: {call.text()}")

# Kind-based search.
calls = root.find_all(kind="call")
print(f"{len(calls)} call expressions total")

# Rule-object search: compose pattern + relational constraints.
inside_fetch = root.find_all(
    {"rule": {"pattern": "print($$$A)", "inside": {"kind": "function_definition"}}}
)
print(f"{len(inside_fetch)} prints inside functions")

# Rewrite: build each replacement from the captured args, then commit the
# edits into a new source string (the original is never mutated).
edits = []
for node in root.find_all(pattern="print($$$ARGS)"):
    args = ", ".join(a.text() for a in node.get_multiple_matches("ARGS"))
    edits.append(node.replace(f"logging.info({args})"))
print(root.commit_edits(edits))

File: examples/node/01-parse-and-rewrite.mjs
// Structural parse, search, and edit with @hyper-light/vorpal-node.
//
//   npm install @hyper-light/vorpal-node
import { parse, Lang } from '@hyper-light/vorpal-node'

const code = `
function fetchUser(id) {
  console.log("fetching", id)
  const res = await client.get('/users/' + id)
  console.log(res.status)
  return res
}
`

const root = parse(Lang.JavaScript, code).root()

// Pattern search with captures.
for (const call of root.findAll('console.log($$$ARGS)')) {
  const { line } = call.range().start
  console.log(`console.log at line ${line + 1}: ${call.text()}`)
}

// Rule-object search: pattern + relational constraint.
const insideFn = root.findAll({
  rule: { pattern: 'console.log($$$A)', inside: { kind: 'function_declaration' } },
})
console.log(`${insideFn.length} logs inside functions`)

// Rewrite: build replacements from captures, commit into a new string.
const edits = root.findAll('console.log($$$ARGS)').map(node => {
  const args = node.getMultipleMatches('ARGS').map(a => a.text()).join(', ')
  return node.replace(`logger.info(${args})`)
})
console.log(root.commitEdits(edits))
